core/domain/src/main/java/com/pronoidsoftware/core/domain/util/Result.kt

```diff
@@ -0,0 +1,20 @@
+package com.pronoidsoftware.core.domain.util
+
+sealed interface Result<out D, out E : Error> {
+    data class Success<out D>(val data: D) : Result<D, Nothing>
+    data class Error<out E : com.pronoidsoftware.core.domain.util.Error>(val error: E) :
+        Result<Nothing, E>
+}
+
+inline fun <T, E : Error, R> Result<T, E>.map(map: (T) -> R): Result<R, E> {
+    return when (this) {
+        is Result.Error -> Result.Error(error)
+        is Result.Success -> Result.Success(map(data))
+    }
+}
+
+fun <T, E : Error> Result<T, E>.asEmptyDataResult(): EmptyResult<E> {
+    return map { }
+}
+
+typealias EmptyResult<E> = Result<Unit, E>
```

Comment:
I recently started adding these two utility functions and I love them:
```
inline fun <T, E: Error> Result<T, E>.onSuccess(action: (T) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> this
        is Result.Success -> {
            action(data)
            this
        }
    }
}
inline fun <T, E: Error> Result<T, E>.onError(action: (E) -> Unit): Result<T, E> {
    return when(this) {
        is Result.Error -> {
            action(error)
            this
        }
        is Result.Success -> this
    }
}
```
Feels much less annoying than having to do when-checks all the time:
```
repository.login(email, password).onSuccess { userData ->
    
}.onError { error ->
    
}
```